Add pivoting blog

kashifest opened the pull request:
This PR adds pivoting blog.

kashifest commented:
/assign @maelk 
/cc @dhellmann @fmuyassarov

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+---
+title: "Metal3 Introduces Pivoting"
```
Introduction to Metal3 pivoting

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+author: Kashif Nizam Khan
+---
+
+`Metal3` project has introduced pivoting in its CI workflow. The motivation of
```
motivation for

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+---
+
+`Metal3` project has introduced pivoting in its CI workflow. The motivation of
+pivoting is to move or pivot all the objects from the ephemeral/management
```
remote 'or pivot '
pivot is what you are defining

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+
+`Metal3` project has introduced pivoting in its CI workflow. The motivation of
+pivoting is to move or pivot all the objects from the ephemeral/management
+cluster to target cluster. This blog post will briefly introduce the concept
```
to a target

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+## What is Pivoting?
+
+In the context of `Metal3` Provider, Pivoting is the process of moving
+`Cluster-API` and `Metal3` objects from the ephemeral k8s cluster to target
```
to a target

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+cluster. In `Metal3`, this process is performed using the
+[clusterctl](https://cluster-api.sigs.k8s.io/clusterctl/overview.html) tool.
+provided by `Cluster-API`. clusterctl recognizes pivoting as `move`. During the
+pivot process clusterctl pauses any reconciliation on `Cluster-API` objects and
```
reconciliation of

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+provided by `Cluster-API`. clusterctl recognizes pivoting as `move`. During the
+pivot process clusterctl pauses any reconciliation on `Cluster-API` objects and
+this gets propagated to CAPM3 objects as well. Once all the objects are paused,
+the objects are created on the other side on target cluster and deleted on
```
on the target

Xenwar replied:
and deleted from the ephemeral cluster.

fmuyassarov commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+In the context of `Metal3` Provider, Pivoting is the process of moving
+`Cluster-API` and `Metal3` objects from the ephemeral k8s cluster to target
+cluster. In `Metal3`, this process is performed using the
+[clusterctl](https://cluster-api.sigs.k8s.io/clusterctl/overview.html) tool.
```
```suggestion
[clusterctl](https://cluster-api.sigs.k8s.io/clusterctl/overview.html) tool
```

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+## Prerequisites
+Prior to the actual pivot process,target cluster should already have the
+provider components, ironic containers and CNI installed and running. To
+be exact the following points need to be addressed:
```
'To be exact ' => Specifically, the following points ...

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+## Pivoting workflow in CI
+The `Metal3` CI currently includes the pivoting as part of the deployment process
+both for `Ubuntu` and `CentOS` based jobs. This essentially means all the PRs
+that go in, are tested through the pivoting workflow also. Here is the CI
```
workflow also => remove also

fmuyassarov commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+[clusterctl](https://cluster-api.sigs.k8s.io/clusterctl/overview.html) tool.
+provided by `Cluster-API`. clusterctl recognizes pivoting as `move`. During the
+pivot process clusterctl pauses any reconciliation on `Cluster-API` objects and
+this gets propagated to CAPM3 objects as well. Once all the objects are paused,
```
```suggestion
this gets propagated to Cluster-api-provider-metal3 (CAPM3) objects as well. Once all the objects are paused,
```

fmuyassarov commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+the objects are created on the other side on target cluster and deleted on
+the ephemeral cluster.
+
+## Prerequisites
+Prior to the actual pivot process,target cluster should already have the
```
```suggestion
Prior to the actual pivot process, target cluster should already have the
```

fmuyassarov replied:
```suggestion
## Prerequisites

Prior to the actual pivot process,target cluster should already have the
```

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+* `Move` all the objects from ephemeral to target cluster.
+* Check the status of the objects to verify whether the objects are being
+reconciled correctly by the controllers in the target cluster. This step
+verifies and finalizes the `pivoting` step.
```
the `pivoting` process.

fmuyassarov commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+provider components, ironic containers and CNI installed and running. To
+be exact the following points need to be addressed:
+
+* `clusterctl` is used to initialize both the bootstrap and target cluster.
```
Maybe instead of bootstrap use either management or ephemeral, to not confuse readers ?

fmuyassarov commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+* `clusterctl` is used to initialize both the bootstrap and target cluster.
+* BMH objects have correct status annotation.
+* Maintain connectivity towards provisioning network.
+* BMO is deployed as part of CAPM3
```
```suggestion
* BMO is deployed as part of CAPM3.
```

fmuyassarov replied:
```suggestion
* Baremetal Operator (BMO) is deployed as part of CAPM3
```

Xenwar commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+and [move_back.yml](https://github.com/metal3-io/metal3-dev-env/blob/master/vm-setup/roles/v1aX_integration_test/tasks/move_back.yml). Under the hood, `pivoting` uses the `move` command from [clusterctl](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html)
+provided by `Cluster-API`.
+
+As stated earlier, all the PRs that go in to any ``Metal3`` repository, the code
```
the first sentence is long and ambiguous.
please break and re-write it.

fmuyassarov commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+`ironic` deployment/containers in the ephemeral cluster. Since we do
+not delete the provider components in the ephemeral cluster,
+installing them again is not necessary.
+* `De-provision` the bmhs and delete the target cluster.
```
```suggestion
* `De-provision` the BMHs and delete the target cluster.
```

fmuyassarov commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+For a detailed explanation of the above mentioned prerequisites please read the
+[pivoting documentation](https://github.com/metal3-io/metal3-docs/blob/master/docs/move.md).  
+
+## Pivoting workflow in CI
+The `Metal3` CI currently includes the pivoting as part of the deployment process
```
```suggestion
## Pivoting workflow in CI

The `Metal3` CI currently includes the pivoting as part of the deployment process
```

fmuyassarov commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,82 @@
+As stated earlier, all the PRs that go in to any ``Metal3`` repository, the code
+change introduced in the PR is verified with pivoting also in the integration
+tests now. Moreover, the upgrade workflow in `Metal3` performs all the upgrade
+operations in `Metal3` after the pivot in the target cluster.
```
```suggestion
operations in `Metal3` after the pivot to the target cluster.
```

fmuyassarov reviewed (commented):
Looks good. Some small nits:

Also, I think it would be better to avoid adding quotes around the word Metal3, Cluster API, Ironic, because when .md file is rendered it looks a bit different on the website. But feel free to keep it if you like so :).

kashifest commented:
@Xenwar @fmuyassarov thanks for the review. Addressed most of them. Kept a few where I felt its subjective, as it is.

maelk commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,85 @@
+the PRs that go in, are tested through the pivoting workflow. Here is the
+CI deployment workflow:
+
+* `make` the [metal3-dev-env](https://github.com/metal3-io/metal3-dev-env.git).
+It gives us the ephemeral cluster with all the  necessary controllers running
+within it.
+* `provision` target cluster. For a normal integration tests, this step deploys
+a control-plane node and a worker in the target cluster. For, `feature-test`
+and `feature-test-upgrade` the provision step deploys three control-planes and a
+worker.
+* Initialize the provider components on the target cluster. This installs all
+the controllers and associated components related to cluster-api ,
+cluster-api-provider-metal3, baremetal-operator and ironic. Since it is
+necessary to have only one set of ironic deployment/containers in the picture,
+this step also deletes the ironic deployment/containers from
+ephemeral cluster.
+* `Move` all the objects from ephemeral to target cluster.
+* Check the status of the objects to verify whether the objects are being
+reconciled correctly by the controllers in the target cluster. This step
+verifies and finalizes the pivoting process.
+* `Move` the objects back to the ephemeral cluster. This step also
+removes the ironic deployment from target cluster and reinstates the
+ironic deployment/containers in the ephemeral cluster. Since we do
+not delete the provider components in the ephemeral cluster,
+installing them again is not necessary.
+* `De-provision` the BMHs and delete the target cluster.
```
Some here are commands "make", others concepts. It would be great if for each step an example command could be given (through makefile or by calling the ansible playbooks)

maelk commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,85 @@
+`move` command from [clusterctl](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html)
+provided by Cluster-API.
+
+As stated earlier, all the PRs that go in to any Metal3 repository, the code
```
for all ?

maelk commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,85 @@
+As stated earlier, all the PRs that go in to any Metal3 repository, the code
+change introduced in the PR is verified with pivoting also in the integration
+tests now. Moreover, the upgrade workflow in Metal3 performs all the upgrade
+operations in Metal3 after the pivot to the target cluster.
```
pivoting ?

maelk reviewed (commented):
Looks good, but could we add more focus on the end-user view of the pivoting feature ? I think there should be a paragraph explaining how to do pivoting outside of Metal3-dev-env because that's what most Metal3 users would want to do. It could link to the docs that we already have a bout pivoting

kashifest commented:
> Looks good, but could we add more focus on the end-user view of the pivoting feature ? I think there should be a paragraph explaining how to do pivoting outside of Metal3-dev-env because that's what most Metal3 users would want to do. It could link to the docs that we already have a bout pivoting

Thanks @maelk for the review, I have addressed your comments. Regarding outside m3-dev-env use, actually the prerequisite paragraph was meant for that and a link to our detailed pivoting doc was already there, I have added another line to be more specific about this in the same paragraph.

maelk commented:
there seem to be some issues to fix (check Travis output)

fmuyassarov commented:
> there seem to be some issues to fix (check Travis output)

@kashifest the issue with Travis should be fixed now in https://github.com/metal3-io/metal3-io.github.io/pull/199. I think you will need to retrigger it.

maelk commented:
there was a problem with the yaspeller. Some words were not added to the dictionary (like PRs, BMHs, Kashif)

kashifest commented:
/retest

kashifest commented:
/re-test

codificat commented:
The link checker is complaining about a (legitimate although unrelated to this PR) broken link: this page does not exist:

https://github.com/openshift-metal3/dev-scripts/blob/master/link-machine-and-node.sh

linked from `_posts/2019-06-25-Metal3.md`.

One option is to replace that link on that file to point to the version available when the post was written, e.g.

https://github.com/openshift-metal3/dev-scripts/blob/1b6768ffeef7f9d3df2f5bdfae5c6829f107ca9c/link-machine-and-node.sh

I know it's not directly related to your post / this PR, but can you add another commit here with the above fix? This should let CI go through (and fix the broken link in the process)

fmuyassarov commented:
@kashifest Can you please rebase your changes, Travis issue should be fixed now.

kashifest commented:
@maelk @fmuyassarov can we take this in? This has been hanging for a while now.
/cc @hardys @russellb

fmuyassarov commented on _posts/2020-12-07-Pivoting.md:
```diff
@@ -0,0 +1,117 @@
+---
+title: "Metal3 Introduces Pivoting"
+date: 2020-12-07
```
We should probably update the date?

kashifest replied:
Done

fmuyassarov reviewed (commented):
LGTM, just minor nit

maelk commented:
/approved

maelk commented:
/approve

maelk commented:
/lgtm

kashifest commented:
/assign @dhellmann

dhellmann commented:
/approve
